<a href="https://colab.research.google.com/github/AlukweJonesTerah/ALGOS/blob/main/Sensor_Data_AFRICA_Air_Quality_Archive_Kisumu_November2018.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("november_2018_sensor_data_archive.csv")

In [4]:
df.head()

sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value
0  122;SDS011;56;-0.108;34.749;2018-11-24T10:49:0...               
1  122;SDS011;56;-0.108;34.749;2018-11-24T10:49:0...               
2  121;DHT22;56;-0.108;34.749;2018-11-24T10:49:07...               
3  121;DHT22;56;-0.108;34.749;2018-11-24T10:49:07...               
4  122;SDS011;56;-0.108;34.749;2018-11-24T10:52:2...

In [6]:
df.columns

Index(['sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value'], dtype='object')

The structure of the data/dataframe should be

1. sensor_id - 122;
2. sensor_type - SDS011;
3. location - 56;
4. lat - -0.108;
5. lon - 34.749;
6. timestamp - 2018-11-24T10:49:06.216079+00:00;
7. value_type - P2;
8. value - 8.65

I will make a new dataframe with each column aligning to each value as indicated above  

## Problems to face there might be empty rows

In [9]:
df.shape

(66574, 1)

In [11]:
df.isnull().sum()

sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value    0
dtype: int64

In [15]:
new_columns = ["sensor_id", "sensor_type", "location", "lat", "lon", "timestamp", "value_type", "value"]
df[new_columns] = df['sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value'].str.split(";", expand=True)

In [18]:
df.drop("sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value", axis=1, inplace=True)

In [21]:
df.shape

(66574, 8)

In [22]:
df.isnull().sum()

sensor_id      0
sensor_type    0
location       0
lat            0
lon            0
timestamp      0
value_type     0
value          0
dtype: int64

## Task 1 is done
we have a new dataframe of shape (66574, 8) and 0 nulls

next we need to check if all raws are with the correct types.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66574 entries, 0 to 66573
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sensor_id    66574 non-null  object
 1   sensor_type  66574 non-null  object
 2   location     66574 non-null  object
 3   lat          66574 non-null  object
 4   lon          66574 non-null  object
 5   timestamp    66574 non-null  object
 6   value_type   66574 non-null  object
 7   value        66574 non-null  object
dtypes: object(8)
memory usage: 4.1+ MB


In [24]:
df.head()

sensor_id sensor_type location     lat     lon  \
0       122      SDS011       56  -0.108  34.749   
1       122      SDS011       56  -0.108  34.749   
2       121       DHT22       56  -0.108  34.749   
3       121       DHT22       56  -0.108  34.749   
4       122      SDS011       56  -0.108  34.749   

                          timestamp   value_type  value  
0  2018-11-24T10:49:06.216079+00:00           P2   8.65  
1  2018-11-24T10:49:06.216079+00:00           P1   9.45  
2  2018-11-24T10:49:07.719863+00:00  temperature  31.70  
3  2018-11-24T10:49:07.719863+00:00     humidity  44.80  
4  2018-11-24T10:52:25.067433+00:00           P2   4.65

In [34]:
print(f' \n Sensor_id unique values: {df["sensor_id"].unique()}')
print(f' \n Senosor_type unique values: {df["sensor_type"].unique()}')
print(f' \n Location unique values: {df["location"].unique()}')
print(f' \n Latitude unique values: {df["lat"].unique()}')
print(f' \n Longitude unique values: {df["lon"].unique()}')
print(f' \n Timestamp unique values: {df["timestamp"].unique()}')
print(f' \n Value types unique values: {df["value_type"].unique()}')
print(f' \n Value unique values: {df["value"].unique()}')

 
 Sensor_id unique values: ['122' '121' '84' '83']
 
 Senosor_type unique values: ['SDS011' 'DHT22']
 
 Location unique values: ['56']
 
 Latitude unique values: ['-0.108']
 
 Longitude unique values: ['34.749']
 
 Timestamp unique values: ['2018-11-24T10:49:06.216079+00:00' '2018-11-24T10:49:07.719863+00:00'
 '2018-11-24T10:52:25.067433+00:00' ... '2018-11-30T23:58:58.909584+00:00'
 '2018-11-30T23:59:34.353157+00:00' '2018-11-30T23:59:35.813678+00:00']
 
 Value types unique values: ['P2' 'P1' 'temperature' 'humidity']
 
 Value unique values: ['8.65' '9.45' '31.70' ... '24.43' '60.75' '53.45']


In [ ]:
df.to_csv("")